# Representação gráfica de funções

Renato Naville Watanabe

## Modelando fenômenos

- Em diferentes áreas da Ciência busca-se modelar fenômenos por médio de funções matemáticas a fim de reproduzir os comportamentos observados na natureza.

- Dado um modelo, muitas vezes, temos a necessidade de visualizar o comportamento do mesmo.

- Gráficos de funções auxiliam o entendimento dos  fenômenos.

![lorentz](https://upload.wikimedia.org/wikipedia/commons/e/ef/Lorenz_Ro28-200px.png)

## Função

- O estudo de funções decorre da necessidade de:
    + Analisar fenômenos, visualizando o comportamento de um sistema.
    + Interpretar interdependências, entendendo como uma variável comporta-se com relação à outra.
    + Encontrar soluções de problemas.
    + Descrever regularidades.
    + Generalizar.

## Definição de uma função

- Uma função é uma regra segundo a qual, para cada elemento $x$ pertencente a um conjunto $A$ corresponde um único elemento $y$ pertencente a um conjunto $B$.

$$y = f(x)$$

- O conjunto $A$ é chamado domínio da função e o conjunto $B$ é o contra-domínio, ou imagem.




## Representação de uma função

- Uma função pode ser representada das seguintes formas: 
    + **Verbalmente** (descrevendo-a com palavras)
    + **Numericamente** (utilizando tabela de valores)
    + **Visualmente** (utilizando gráficos)
    + **Algebricamente** (utilizando uma fórmula explícita)

## Ferramentas de visualização

Existem diversas ferramentas para utilizadas em calculos matemáticos avançados.
- Matlab
- Mathematica
- Octave
- Scilab
- R
- [**Python**](https://nbviewer.jupyter.org/github/BMClab/bmc/blob/master/notebooks/PT-BR/PythonForScientificComputing_PT-BR.ipynb)
- Fortran
- Julia

Geralmente contam com bibliotecas de funções matemáticas prontas e recursos avançados.

## Spyder 
Ambiente de programação Python que permite a criação de novas funções/bibliotecas do usuário.

<img src="imagens/spyder.png" width="400">

O Spyder pode ser obtido com o pacote [Anaconda](https://www.anaconda.com/distribution/).

## Spyder

A interação do usuário com o Spyder pode ocorrer de duas formas distintas:

- Na **primeira forma**, os comando são digitados diretamente no *console* do Spyder:
    + Ao ser pressionada a tecla Enter, os comandos digitados são interpretados e imediatamente executados.
    + O Spyder (Python) funciona como uma  calculadora sofisticada.

- Na **segunda forma**, um conjunto de comandos é digitado em um arquivo texto:
    + Este arquivo  (script) pode ser  executado.
    + Neste modo, o Spyder funciona como um ambiente de programação.

## Spyder

Digitando o seguinte comando no console:

In [4]:
x = 2

estaremos criando uma variável real chamada $x$ cujo valor é igual a 2.

## Variável

Uma variável é um endereço da memória de acesso randômico (RAM), representada por um nome 
(rótulo), criado pelo usuário, cujo conteúdo pode se alterar no decorrer do programa.

Uma variável é composta por dois elementos:
    
  - **Identificador**: nome dado pelo programador à variável
  - **Conteúdo**: valor da variável
<img src="imagens/memvar2.png" width="400">

No caso anterior, x é o identificador (nome da variável) e 2 é o conteúdo (valor).

## Variável

Cada vez que mencionarmos o nome de uma variável, estaremos na verdade utilizando o seu conteúdo

In [16]:
y = x + 5
y

7

Esta operação define $y$ como sendo uma variável com valor igual ao valor de $x$ mais cinco. Ou seja, $y$ tem um valor igual a 7

In [17]:
z = x*y
z

14

Neste caso, $z$ é igual à multiplicação dos valores guardados em $x$ e $y$, ou seja, $z$ é igual a 14.

In [18]:
w = z/x
w

7.0

Aqui, $w$ é igual à divisão dos valores guardados em $z$ e $x$, ou seja, $w$
é igual a 7.

In [19]:
k = w-x
k

5.0

Aqui, $k$ é igual à subtração do valor guardado em $x$ de $w$. Ou seja, $k$
é igual a 5.

In [21]:
m = k**3
m

125.0

Aqui, $m$ é igual ao valor guardado em $k$ elevado ao cubo. Ou seja, $m$ é igual a 125.

## Funções matemáticas

Para usar funções mais complexas, é necessário utilizar a biblioteca Numpy. Para poder utilizar esta biblioteca, é necessário importá-la:

In [25]:
import numpy as np

Esta biblioteca tem diversas funções que são facilmente utilizadas colocando 'np.' antes da função desejada. Por exemplo, a função logaritmo na base 10 é chamada com: 

In [27]:
np.log10(16)

1.2041199826559248

Se você quiser guardar o resultado em uma variável também é possível:

In [30]:
a = np.log10(16)
a

1.2041199826559248

E se quiser calcular o valor da função para uma variável é simples:

In [32]:
b = np.log10(k)
b

0.69897000433601886

## Funções matemáticas

Alguns exemplos de funções são:


**Exponencial**: exp($x$)=$e^x$ em que $e=2,7182...$ 

In [35]:
np.exp(-2)

0.1353352832366127

**Raiz quadrada**

In [37]:
s = np.sqrt(25)
s

5.0

**Funções trigonométricas**

In [39]:
d = np.sin(np.pi/2)
g = np.cos(np.pi/4)
print(d)
print(g)

1.0
0.707106781187


Nos dois últimos casos, para representar o valor de $\pi$, foi utilizado 'np.pi'.

## Vetores

Com o Numpy você pode criar vetores. 
Com o comando 'np.array' você cria vetores digitando cada elemento pertencente ao vetor:

In [41]:
x = np.array([1,2,3,4,5])
x

array([1, 2, 3, 4, 5])

Com o comando 'np.arange', você cria vetores especificando os limites e o intervalo entre os elementos. Por exemplo, para criar um vetor que vai de 0 a 12 com intervalo entre os elementos de 0,2, deve-se digitar:

In [45]:
y = np.arange(0, 12, 0.2)
y

array([  0. ,   0.2,   0.4,   0.6,   0.8,   1. ,   1.2,   1.4,   1.6,
         1.8,   2. ,   2.2,   2.4,   2.6,   2.8,   3. ,   3.2,   3.4,
         3.6,   3.8,   4. ,   4.2,   4.4,   4.6,   4.8,   5. ,   5.2,
         5.4,   5.6,   5.8,   6. ,   6.2,   6.4,   6.6,   6.8,   7. ,
         7.2,   7.4,   7.6,   7.8,   8. ,   8.2,   8.4,   8.6,   8.8,
         9. ,   9.2,   9.4,   9.6,   9.8,  10. ,  10.2,  10.4,  10.6,
        10.8,  11. ,  11.2,  11.4,  11.6,  11.8])

Com o comando 'np.linspace', você cria vetores especificando os limites e o número de elementos. Por exemplo, para criar um vetor que vai de 0 a 11 com 130 elementos, deve-se digitar:

In [46]:
z = np.linspace(0, 11, 130)
z

array([  0.        ,   0.08527132,   0.17054264,   0.25581395,
         0.34108527,   0.42635659,   0.51162791,   0.59689922,
         0.68217054,   0.76744186,   0.85271318,   0.9379845 ,
         1.02325581,   1.10852713,   1.19379845,   1.27906977,
         1.36434109,   1.4496124 ,   1.53488372,   1.62015504,
         1.70542636,   1.79069767,   1.87596899,   1.96124031,
         2.04651163,   2.13178295,   2.21705426,   2.30232558,
         2.3875969 ,   2.47286822,   2.55813953,   2.64341085,
         2.72868217,   2.81395349,   2.89922481,   2.98449612,
         3.06976744,   3.15503876,   3.24031008,   3.3255814 ,
         3.41085271,   3.49612403,   3.58139535,   3.66666667,
         3.75193798,   3.8372093 ,   3.92248062,   4.00775194,
         4.09302326,   4.17829457,   4.26356589,   4.34883721,
         4.43410853,   4.51937984,   4.60465116,   4.68992248,
         4.7751938 ,   4.86046512,   4.94573643,   5.03100775,
         5.11627907,   5.20155039,   5.28682171,   5.37

## Gráfico de funções

Para fazer o gráfico de funções no Python, deve-se importar a biblioteca matplotlib.

In [47]:
import matplotlib.pyplot as plt

## Referências

- Chalco, JM, *Slides de Bases Computacionais da Ciência*, (2014)
- Leite, S, *Slides de Bases Computacionais da Ciência*, (2018)
- [Marietto, MGB et al.; **Bases computacionais da Ciência** (2013)](http://prograd.ufabc.edu.br/images/pdf/bases_computacionais_livro.pdf).
- [Duarte, M, *Python para computação científica* (2017)](https://nbviewer.jupyter.org/github/BMClab/bmc/blob/master/notebooks/PT-BR/PythonForScientificComputing_PT-BR.ipynb)